### PR TITLE
struct array lowering to __equals template call

### PR DESCRIPTION
**Pull request by somzzz**
Complement of https://github.com/dlang/druntime/pull/1792

### CODE REVIEW

File: src/ddmd/expression.d [hunk 1]
```diff
@@ -15669,6 +15670,38 @@ extern (C++) final class EqualExp : BinExp
...
 
+        Type t1 = e1.type.toBasetype();
+        Type t2 = e2.type.toBasetype();
+
+        // Lower array of struct comparisons to druntime __equals
+        if ((t1.ty == Tarray || t1.ty == Tsarray) &&
+            (t2.ty == Tarray || t2.ty == Tsarray))
+        {
+            Type telement  = t1.nextOf().toBasetype();
+            Type telement2 = t2.nextOf().toBasetype();
```
Comment (andralex) [inline src/ddmd/expression.d]: Shall we use the naming convention with `lhs` and `rhs` here as well?

**Review by andralex — APPROVED** [review]
LGTM. @WalterBright ?

**Comment by quickfur** [comment]
Awesome!

File: src/ddmd/expression.d (hunk 1 above)
Comment (WalterBright) [inline src/ddmd/expression.d]: '1' and '2' are used throughought

**Review by WalterBright — APPROVED** [review]

**Comment by somzzz** [comment]
I extracted the failing code from that test:
```
string op;
struct S
{
    char x = 'x';
    this(this) { op ~= x-0x20; }    // upper case
    ~this()    { op ~= x; }         // lower case
}
void main()
{
    [S('b'), S('c')] == [S('b'), S('c')];
}
```
Throws: ```core.exception.InvalidMemoryOperationError@src/core/exception.d(696): Invalid memory operation```

1. It has nothing to do with the code for ```__equals``` in ```druntime``` which I entirely commented.
2. It doesn't throw if I do this in main:
```
    S[3] sa = [S('a'), S('b'), S('c')];
    sa[1..3] == sa[1..3];
```
3. It doesn't throw if I don't do anything in the ctor and dtor:
```
string op;
struct S
{
    char x = 'x';
    this(this) {  }  // this can be entirely removed
    ~this()    { }  // same here
}
void main()
{
    [S('b'), S('c')] == [S('b'), S('c')];
}
```

So it seems that altering ```op``` in the struct ```ctor/dtor``` for arrays of structs which are not assigned to any variable causes the error. I'm not sure how this affects the current ```dmd``` patch and how this should be fixed.

**Comment by ZombineDev** [comment]
The reason for the `InvalidMemoryOperationError` exception is that the code is trying to use the GC while the GC is performing a collection. Usually this happens when:
1. You have a heap allocated class or struct, e.g. `S`
2. A GC collection is triggered
3. The GC finds that `S` is no longer referenced by anyone and calls its destructor prior to deallocating its memory
4. `S.~this()` causes a GC allocation by appending to `op`
5. The GC throws `InvalidMemoryOperationError`

The case posted above (which is a bit different than the one in `runnable/sdtor.d`) will always fail (proof: https://dpaste.dzfl.pl/cce656d4aeb6) regardless of your change, due to the nature of D's GC.

The reason the test case didn't fail before is that it was written carefully to avoid `S[]` heap allocations by using static arrays. I can't tell by looking at your change if there's anything wrong with it, but it seams that something now forces the array literals to go on the heap. My guess is that the compiler recognizes a limited number of cases where it is able to elide heap allocations and now the new AST doesn't match its heuristics...

... confirmed!:
https://dpaste.dzfl.pl/77f44eb3448b
```
void main()
{
	test1();
	test2(); // doesn't compile
}

void test1()() nothrow @nogc
{
	int[3] arr = [1, 2, 3];
	
	assert (arr == [1, 2, 3]);
}

void test2()() nothrow @nogc
{
	int[3] arr = [1, 2, 3];
	
	check(arr, [1, 2, 3]);
}

void check(int[] lhs, int[] rhs)
{
	assert (lhs == rhs);
}
```

**Comment by ZombineDev** [comment]
In summary issue is that the compiler treats `assert (arr == [1, 2, 3])` different from `check(arr, [1, 2, 3])` (and even `(lhs, rhs){ assert (lhs == rhs); }(arr, [1, 2, 3])`). The simple `assert` expression doesn't cause a GC allocation, while the other ones do.
I see several ways forward:
1. Workaround the issue in `runnable/sdtor.d`. A hack would probably be okish in the short term, but your change breaks `@nogc` code that assumes that comparison with literals won't allocate. The same is probably true for `__cmp(T)(const T[] lhs, const T[] rhs)`.
2. Add a special case for built-in ops like `__cmp` and `__equals`
3. Improve the GC allocation elision algorithm in the compiler. This would be great because it will benefit existing user code, but it is also more difficult and time-consuming effort.

**Comment by andralex** [comment]
Why did the bot close this?

**Pull request closed without merging**